### ISSUE TITLE
Individualize SQL Query Threading

### ISSUE THREAD

**Issue by ttasdasda** [alliedmodders/sourcemod#1076]
# Help us help you
  - [x] I have checked that my issue [doesn't exist yet](https://github.com/alliedmodders/sourcemod/issues).
  - [x] I have tried my absolute best to reduce the problem-space and have provided the absolute smallest test-case possible.

# Description
As time goes on, plugins become more feature-rich and the databases get bigger (sometimes even bloated; we're living in the age of big data, after all), and single-threaded queries no longer cut it in certain use-cases. Multiple-join queries can take several seconds to process, and running several of them consequently (by mere coincidence or someone's malicious intent) can pretty much render the server inoperable - a player on a surf-server wouldn't be able to finish his run (an almost-instant insert query) while some other player is loading some personal data (a 5-sec select query).

Even spreading the queries across different databases changes nothing: a query to DB #1 will inevitably delay a consequent query to DB #2.
At the start of the map, it takes over 30 seconds just to load all the necessary data. Doesn't help that there's an arbitrary amount of overhead added to fast queries because of DEFAULT_THINK_TIME_MS.

I understand this is sort of a niche problem, as most smaller/non-surf servers don't really need it and would probably laugh it off, but for us, it's really murdering the player experience. This is a long-awaited feature.

**Comment by KyleSanderson** [comment]
I agree - this is important to have and has long been on a (frankly) my roadmap to work on.

In the meantime can you try using threaded queries? It should get you out of this rut. The pattern I personally use is OnMapEnd to grab the new data required (and if it's not critical - make it a low priority) and update my cached trie data only when I have the new results.

**Comment by dvander** [comment]
I'm confused, doesn't SQL multi-threading already exist via the async API? The synchronous API should not be used, pretty much ever.

**Comment by KyleSanderson** [comment]
> I'm confused, doesn't SQL multi-threading already exist via the async API? The synchronous API should not be used, pretty much ever.

From my understanding - they're saying their SQL server is incredibly slow and not tuned properly. This is partially on us as we're using a single pipeline for all DBI transactions (SQLite and MySQL for instance are shared on the same thread). I had severe problems when SQL_TConnect was timing out as everything else was being blocked by the "high" priority traffic (my sysadmin thumb'd an entry in databases.cfg).

The request is an enhancement, maybe minor defect with the implementation of the time but my word the ecosystem was very different a decade ago.

**Comment by gonzalopixart** [comment]
sourcemod not have async calls and await?

**Comment by dvander** [comment]
No, it does not. Note that Callbacks returning values or Actions are inherently not awaitable. How could we pause and resume OnClientConnect, or command callbacks? We can't, so those functions are fundamentally unable to "await" without simply dropping the async return value. This would be a source of confusion.

The other source of confusion is that games are synchronous state machines. Consider the following hypothetical code:

[code]async Action TimerCallback(Handle:timer, data)
{
    int client = GetClientOfUserId(data);
    char auth[64];
    GetClientAuthString(client, auth, sizeof(auth));

    await CheckClientInDatabase(auth);

    PrintToChat(client, "Congrats, you're in the database!");
}[/code]

In between the "await" and "Print", it is possible for any number of things to happen - "client" may have since been disconnected, and a new client could have joined. Another instance of the exact same timer could have already run and completed. So, any game state from before the "await" must be treated as invalid, which is a burden that must be suffered regardless of whether promises or callbacks are used. In fact, not having lambdas or variable captures makes this very explicit.

Even if we solved every other technical problem around async/await, I would not want to introduce it unless we were able to systematically address the issues above.

**Comment by gonzalopixart** [comment]
I understand perfectly, that happens in many environments today with doing asynchronous tasks, such as in javascript that currently simulated it with promises, but there is a very interesting fact in this technology that the use has to be precise for certain occasions such as in your example n would be useful to implement it, and if in sql requests. I understand that today you use a single source mod thread with the engine, right? Do you not have the possibility of directing in a new thread another type of asynchronous requests and joining them when necessary in the thread that is sourcemod?

> No, no lo hace. Tenga en cuenta que las devoluciones de llamada que devuelven valores o acciones no son inherentemente esperadas. ¿Cómo podríamos pausar y reanudar OnClientConnect, o ordenar devoluciones de llamada? No podemos, por lo que esas funciones son fundamentalmente incapaces de "esperar" sin simplemente soltar el valor de retorno asíncrono. Esto sería una fuente de confusión.
> 
> La otra fuente de confusión es que los juegos son máquinas de estado sincrónicas. Considere el siguiente código hipotético:
> 
> ```
> async Action TimerCallback(Handle:timer, data)
> {
>     int client = GetClientOfUserId(data);
>     char auth[64];
>     GetClientAuthString(client, auth, sizeof(auth));
> 
>     await CheckClientInDatabase(auth);
> 
>     PrintToChat(client, "Congrats, you're in the database!");
> }
> ```
> 
> Entre "esperar" e "Imprimir", es posible que suceda cualquier cantidad de cosas: "cliente" puede haber sido desconectado y un nuevo cliente podría haberse unido. Otra instancia del mismo temporizador ya podría haberse ejecutado y completado. Por lo tanto, cualquier estado del juego anterior al "espera" debe ser tratado como inválido, lo cual es una carga que debe ser soportada independientemente de si se usan promesas o devoluciones de llamada. De hecho, no tener lambdas o capturas variables lo hace muy explícito.
> 
> Incluso si resolviéramos cualquier otro problema técnico relacionado con async / wait, no me gustaría presentarlo a menos que pudiéramos abordar sistemáticamente los problemas anteriores.

I understand perfectly, that happens in many environments today with doing asynchronous tasks, such as in javascript that currently simulated it with promises, but there is a very interesting fact in this technology that the use has to be precise for certain occasions such as in your example n would be useful to implement it, and if in sql requests. I understand that today you use a single source mod thread with the engine, right? Do you not have the possibility of directing in a new thread another type of asynchronous requests and joining them when necessary in the thread that is sourcemod?
